### ISSUE TITLE
Overload in runtime

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#12077]
Migrated from [rt.perl.org#112708](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=112708) (status was 'resolved')

Searchable as RT112708$

**Comment by p5pRT** [comment]
### From @tokuhirom

#### Created by tokuhirom@gmail.com



&nbsp;  > use strict;
&nbsp;  > use warnings;
&nbsp;  > use utf8;
&nbsp;  > use 5\.010000;
&nbsp;  >
&nbsp;  > \{
&nbsp;  >     package Foo;
&nbsp;  >     sub new \{ bless \+\{\}\, shift \}
&nbsp;  > \}
&nbsp;  >
&nbsp;  > my $x = Foo\->new\(\);
&nbsp;  > say\($x\);
&nbsp;  >
&nbsp;  > \{
&nbsp;  >     eval q\!
&nbsp;  >     package Foo;
&nbsp;  >     use overload q\{""\} => sub \{ "STRINGIFY" \};
&nbsp;  >     \!;
&nbsp;  >     die $@&#8203; if $@&#8203;;
&nbsp;  > \}
&nbsp;  >
&nbsp;  > say\($x\);
&nbsp;  > my $y = Foo\->new\(\);
&nbsp;  > say\($x\);
&nbsp;  > say\($y\);

And results are

&nbsp;  > Foo=HASH\(0x1d8ab78\)
&nbsp;  > Foo=HASH\(0x1d8ab78\)
&nbsp;  > Foo=HASH\(0x1d8ab78\)
&nbsp;  > STRINGIFY

overloaded operators are applied to values create after overload only?

I seems it's a bug\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl 5.14.2:

Configured by Debian Project at Fri Mar 23 17:14:02 UTC 2012.

Summary of my perl5 (revision 5 version 14 subversion 2) configuration:
   
  Platform:
    osname=linux, osvers=2.6.24-31-server, archname=x86_64-linux-gnu-thread-multi
    uname='linux allspice 2.6.24-31-server #1 smp tue feb 14 13:12:22 utc 2012 x86_64 x86_64 x86_64 gnulinux '
    config_args='-Dusethreads -Duselargefiles -Dccflags=-DDEBIAN -Dcccdlflags=-fPIC -Darchname=x86_64-linux-gnu -Dprefix=/usr -Dprivlib=/usr/share/perl/5.14 -Darchlib=/usr/lib/perl/5.14 -Dvendorprefix=/usr -Dvendorlib=/usr/share/perl5 -Dvendorarch=/usr/lib/perl5 -Dsiteprefix=/usr/local -Dsitelib=/usr/local/share/perl/5.14.2 -Dsitearch=/usr/local/lib/perl/5.14.2 -Dman1dir=/usr/share/man/man1 -Dman3dir=/usr/share/man/man3 -Dsiteman1dir=/usr/local/man/man1 -Dsiteman3dir=/usr/local/man/man3 -Duse64bitint -Dman1ext=1 -Dman3ext=3perl -Dpager=/usr/bin/sensible-pager -Uafs -Ud_csh -Ud_ualarm -Uusesfio -Uusenm -Ui_libutil -DDEBUGGING=-g -Doptimize=-O2 -Duseshrplib -Dlibperl=libperl.so.5.14.2 -des'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DDEBIAN -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2 -g',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DDEBIAN -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.6.3', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib/x86_64-linux-gnu /lib/../lib /usr/lib/x86_64-linux-gnu /usr/lib/../lib /lib /usr/lib
    libs=-lgdbm -lgdbm_compat -ldb -ldl -lm -lpthread -lc -lcrypt
    perllibs=-ldl -lm -lpthread -lc -lcrypt
    libc=, so=so, useshrplib=true, libperl=libperl.so.5.14.2
    gnulibc_version='2.15'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -g -L/usr/local/lib -fstack-protector'

Locally applied patches:
    


@INC for perl 5.14.2:
    /etc/perl
    /usr/local/lib/perl/5.14.2
    /usr/local/share/perl/5.14.2
    /usr/lib/perl5
    /usr/share/perl5
    /usr/lib/perl/5.14
    /usr/share/perl/5.14
    /usr/local/lib/site_perl
    .


Environment for perl 5.14.2:
    HOME=/home/tokuhirom
    LANG=en_US.UTF-8
    LANGUAGE (unset)
    LC_ALL=C
    LC_CTYPE=C
    LC_DATE=C
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/tokuhirom/perl5/perlbrew/bin:/usr/local/bin/:/usr/local/bin/:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/bin/X11:/usr/games
    PERLBREW_BASHRC_VERSION=0.42
    PERLBREW_HOME=/home/tokuhirom/.perlbrew
    PERLBREW_PATH=/home/tokuhirom/perl5/perlbrew/bin
    PERLBREW_ROOT=/home/tokuhirom/perl5/perlbrew
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @cpansprout

On Mon Apr 30 20&#8203;:05&#8203;:36 2012\, tokuhirom wrote&#8203;:
>   > use strict;
>   > use warnings;
>   > use utf8;
>   > use 5\.010000;
>   >
>   > \{
>   >     package Foo;
>   >     sub new \{ bless \+\{\}\, shift \}
>   > \}
>   >
>   > my $x = Foo\->new\(\);
>   > say\($x\);
>   >
>   > \{
>   >     eval q\!
>   >     package Foo;
>   >     use overload q\{""\} => sub \{ "STRINGIFY" \};
>   >     \!;
>   >     die $@&#8203; if $@&#8203;;
>   > \}
>   >
>   > say\($x\);
>   > my $y = Foo\->new\(\);
>   > say\($x\);
>   > say\($y\);
> 
> And results are
> 
>   > Foo=HASH\(0x1d8ab78\)
>   > Foo=HASH\(0x1d8ab78\)
>   > Foo=HASH\(0x1d8ab78\)
>   > STRINGIFY
> 
> overloaded operators are applied to values create after overload only?
> 
> I seems it's a bug\.

Internally\, for efficiency’s sake\, the object is flagged as having
overloading \(SvAMAGIC\) so that code that might handle overloading can
check quickly\.

The problem here is that\, when the overloadedness of the package
changes\, there is no list of its objects anywhere\, so they cannot all be
updated\.

Would it be possible to move the AMAGIC flag from the object to its
stash?  Instead of SvFLAGS\(SvRV\(sv\)\) & SVf\_AMAGIC\, we would have
SvFLAGS\(SvSTASH\(SvRV\(sv\)\)\) & SVf\_AMAGIC\, which would not be noticeably
slower\.

We could have the overloadedness calculated by Perl\_Gv\_AMupdate when
‘use overload’ finishes\, rather than when bless happens\.

Is there anything wrong with that?

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @ikegami

On Thu\, May 17\, 2012 at 3&#8203;:53 PM\, Father Chrysostomos via RT \<
perlbug\-followup@&#8203;perl\.org> wrote&#8203;:

> Would it be possible to move the AMAGIC flag from the object to its
> stash?  Instead of SvFLAGS\(SvRV\(sv\)\) & SVf\_AMAGIC\, we would have
> SvFLAGS\(SvSTASH\(SvRV\(sv\)\)\) & SVf\_AMAGIC\, which would not be noticeably
> slower\.
>

Overloaded operators are inherited\, so that doesn't sound like enough?


package ClassA \{
&nbsp;   use overload '""' => sub \{ "stringify" \};
&nbsp;   sub new \{ bless\(\{\}\, shift\) \}
\}

package ClassB \{
&nbsp;   our @&#8203;ISA = 'ClassA';
\}

print ClassB\->new\(\)\, "\\n";  \# stringify

**Comment by p5pRT** [comment]
### From @cpansprout

On Thu May 17 15&#8203;:56&#8203;:12 2012\, ikegami@&#8203;adaelis\.com wrote&#8203;:
> On Thu\, May 17\, 2012 at 3&#8203;:53 PM\, Father Chrysostomos via RT \<
> perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> 
> > Would it be possible to move the AMAGIC flag from the object to its
> > stash?  Instead of SvFLAGS\(SvRV\(sv\)\) & SVf\_AMAGIC\, we would have
> > SvFLAGS\(SvSTASH\(SvRV\(sv\)\)\) & SVf\_AMAGIC\, which would not be noticeably
> > slower\.
> >
> 
> Overloaded operators are inherited\, so that doesn't sound like enough?
> 
> 
> package ClassA \{
>    use overload '""' => sub \{ "stringify" \};
>    sub new \{ bless\(\{\}\, shift\) \}
> \}
> 
> package ClassB \{
>    our @&#8203;ISA = 'ClassA';
> \}
> 
> print ClassB\->new\(\)\, "\\n";  \# stringify

You’re right\.  That’s not enough\.  Each stash has its own overload
method cache\.  Currently\, those are updated during bless\.  Simply moving
the cache update to ‘use overload’ will shift bugs around\, because
setting @&#8203;ISA or redefining a method should do the same thing\, or at
least mark the cache as stale\.

Here’s another bug that results from this&#8203;:

$ \./perl \-Ilib \-e 'use overload \("abs"\)x2; sub abs \{warn "one"\} $o =
bless\[\]; eval q=sub abs \{ warn "two" \}=; abs $o; $o\->abs'
one at \-e line 1\.
two at \(eval 1\) line 1\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @cpansprout

I have a fix waiting on the sprout/misc\-post\-5\.16 branch\, which\, er\,
somehow ended up 24 commits long\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @cpansprout

Now fixed by commit 234df27b26 and a few commits leading up to it\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
